File: binsV1/binV5/sixth_problem.rb
require 'json'

staffs = JSON.parse(File.read('input_data/staff.json'), symbolize_names: true)

highest_rated_staffs = staffs.select { |staff| staff[:result][:rating] > 4  }

highest_rated_staffs.map! do |staff|
  {
    id: staff[:id],
    full_name: staff[:full_name],
    email: staff[:email],
    rating: staff[:result][:rating]
  }
end

print JSON.pretty_generate({ data: highest_rated_staffs })

File: quizzes/quizzesV1/quizzes3/sixth_problem.js
const fs = require("fs");

const staffs = JSON.parse(fs.readFileSync("input_data/staff.json", "utf8"));

const findTopRatedStaffsWithChained = () => {
  return staffs.flatMap((staff) => {
    if (staff?.result?.rating <= 4) return [];

    return {
      id: staff.id,
      full_name: staff.full_name,
      email: staff.email,
      rating: staff.result.rating,
    };
  });
};

const findTopRatedStaffsWithLoop = () => {
  const topRatedStaffs = [];

  for (let i = 0; i < staffs.length; i++) {
    const staff = staffs[i];

    if (staff?.result?.rating > 4) {
      topRatedStaffs.push({
        id: staff.id,
        full_name: staff.full_name,
        email: staff.email,
        rating: staff.result.rating,
      });
    }
  }

  return topRatedStaffs;
};

console.log(JSON.stringify({ data: findTopRatedStaffsWithChained() }, null, 2));
console.log(JSON.stringify({ data: findTopRatedStaffsWithLoop() }, null, 2));
